Mathlib/LinearAlgebra/Matrix/SVD/Defs.lean

```diff
@@ -0,0 +1,479 @@
+      (hAHA.eigenvectorMatrixᴴ)) _ (eigenColumnEquiv A).symm _,
+    ← submatrix_mul_equiv (diagonal (IsROrC.ofReal ∘ hAHA.eigenvalues)) (hAHA.eigenvectorMatrixᴴ)
+    _ (eigenColumnEquiv A).symm _,
+    ← @IsROrC.algebraMap_eq_ofReal 𝕂]
+  simp_rw [Function.comp]
+  rw [← diagonal_map, submatrix_map, ← reindex_apply, ← Equiv.coe_refl, ← Equiv.refl_symm,
+    ← reindex_apply, ←conjTranspose_submatrix, ← reindex_apply, S_block, V_columns,
+    conjTranspose_fromColumns_eq_fromRows_conjTranspose, fromBlocks_map, fromBlocks_mul_fromRows,
+    fromColumns_mul_fromRows]
+  simp only [map_zero, Matrix.map_zero, Matrix.zero_mul, add_zero, Matrix.mul_zero]
+  rw [Matrix.mul_assoc]
+  apply map_zero
+
+lemma reduced_spectral_theorem' (A : Matrix (Fin M) (Fin N) 𝕂) :
+    A  *  Aᴴ = A.svdU₁'  *  (A.svdμ'.map (algebraMap ℝ 𝕂))  *  A.svdU₁'ᴴ := by
+  let hAAH := isHermitian_mul_conjTranspose_self A
+  rw [← submatrix_id_id (A * Aᴴ), IsHermitian.spectral_theorem' hAAH,
+    ← IsHermitian.conjTranspose_eigenvectorMatrix, Matrix.mul_assoc,
+    ← submatrix_mul_equiv hAAH.eigenvectorMatrix
+      (diagonal (IsROrC.ofReal ∘ hAAH.eigenvalues)  *  (hAAH.eigenvectorMatrixᴴ)) _
+        (eigenRowEquiv A).symm _,
+    ← submatrix_mul_equiv (diagonal (IsROrC.ofReal ∘ hAAH.eigenvalues))
+      (hAAH.eigenvectorMatrixᴴ) _ (eigenRowEquiv A).symm _,
+      ← @IsROrC.algebraMap_eq_ofReal 𝕂]
+  simp_rw [Function.comp]
+  rw [← diagonal_map, submatrix_map,
+    ← reindex_apply, ← Equiv.coe_refl, ← Equiv.refl_symm, ← reindex_apply,
+    ← conjTranspose_submatrix, ← reindex_apply, S'_block, U_columns',
+    conjTranspose_fromColumns_eq_fromRows_conjTranspose, fromBlocks_map,
+    fromBlocks_mul_fromRows, fromColumns_mul_fromRows]
+  simp only [map_zero, Matrix.map_zero, Matrix.zero_mul, add_zero, Matrix.mul_zero]
+  rw [Matrix.mul_assoc]
+  rw [map_zero]
+
+open scoped ComplexOrder
+lemma eig_vals_ne_zero_pos {m n: Type} [Fintype m] [Fintype n] [DecidableEq n]
+    (A : Matrix m n 𝕂) (z: {a // (isHermitian_transpose_mul_self A).eigenvalues a ≠ 0 }) :
+    0 < ((isHermitian_transpose_mul_self A).eigenvalues z) :=
+  lt_of_le_of_ne
+    (Matrix.PosSemidef.eigenvalues_nonneg (posSemidef_conjTranspose_mul_self _) _) -- 0 ≤ _
+    (z.prop.symm) -- 0 ≠ _
+
+lemma svdσ_inv (A : Matrix (Fin M) (Fin N) 𝕂) : A.svdσ⁻¹ =
+    (reindex
+      (finRankEquivEigsConjTransposeMulSelf A).symm
+      (finRankEquivEigsConjTransposeMulSelf A).symm)
+      (diagonal (fun (i : {a // (isHermitian_transpose_mul_self A).eigenvalues a ≠ 0}) =>
+        1 / Real.sqrt ((isHermitian_transpose_mul_self A).eigenvalues i))) := by
+  apply inv_eq_right_inv
+  rw [svdσ]
+  simp only [ne_eq, reindex_apply, submatrix_diagonal_equiv, diagonal_mul_diagonal,
+    Function.comp_apply]
+  rw [← diagonal_one, diagonal_eq_diagonal_iff]
+  intros i
+  rw [mul_one_div_cancel]
+  apply ne_of_gt
+  apply (Real.sqrt_pos.2 (eig_vals_ne_zero_pos _ _))
```

Comment:
For example based on my remark on `eig_vals_ne_zero_pos`, this should simplify to something like (untested):
```suggestion
  apply Real.sqrt_ne_zero.mpr i.prop
```

Reply:
If I understand what you are suggesting I don't think this will work directly as you describe it. The indirect effect of the double reindexing I did is to hide the inner prop (which says the eigenvalue ne zero). Recall that we reindex first from `Fin N` to `{i // eig ~= 0} \oplus {i // eig = 0}` then we reindex again to Fin r \oplus Fin (N - r)
I think the cost of getting it back from under these definitions will be longer than what I am doing now.

Reply:
On trying a bit harder it seems I can access the underlying proof which appears in InfoView as 
```
apply (Real.sqrt_ne_zero _).mpr
apply ((finRankEquivEigsConjTransposeMulSelf A).1 i).prop
```

Reply:
So I guess I was able to implement your suggestion. The difference is now I have this term three times:

```lean
  (Real.sqrt_ne_zero (eigenvalues_conjTranspose_mul_self_nonneg _)).2
      ((finRankEquivEigsConjTransposeMulSelf A).1 i).prop)
```
In terms of your suggestion it is fewer by 2 lines discounting the lemmas that will be moved into PR #7312 
So I will go with it